8.0 add support for rental

alexis-via opened the pull request:
These rental modules were initially developed on https://code.launchpad.net/openerp-rental
So it's now time to move to github/OCA

Commit:
Initial check-in of the rental project.

Commit:
We want to have the date_planned of the regular products that are on the same SO as the rentals = the start date of the rentals, because we suppose that these products are "accessories" of the rental.
Re-organise the order of some lines of code to make it more "logic".

Commit:
Restrict to "day" UoM in the constraint.
Add list of limitations in the module description.

Commit:
Add read access on rental to stock user.

Commit:
Add POT file
Add full FR translation for sale_start_end_dates
Add partial FR translation for sale_rental

Commit:
FIX missing dependancy.

Commit:
Add pt_BR translation, contributed by Caio Barros.

Commit:
[FIX] Missing call to super in _invoice_line_hook
This bug was very annoying because it means that, if you had sale_rental installed, you are missing the link betweek purchase order lines and invoice lines (for purchase orders created from pickings)

Commit:
Better default values in the wizard for the creation of rental services from products

Commit:
Add a note about the fact that you need to have access to the form view of sale order line, and instructions on how to do it
Add demo data to automatically add admin and demo to group sale.group_mrp_properties, to have access to form view of sale order lines.

Commit:
FIX my commit revno 9 on the wizard to create the rental service
Add a link to the screencast in the module description

Commit:
Port to v8 and new API with several enhancements

Commit:
Merge with the 8.0 branch of lp:openerp-rental

Commit:
Fix typo

bealdav commented on sale_rental/__init__.py:
```diff
@@ -0,0 +1,24 @@
+##############################################################################
+
+from . import rental
+from . import wizard
```
line 1 to 22 are not useful, just import no license mandatory, no explicit encoding, just ascii used
It's just a suggestion, you may keep as it if you want

Commit:
Add account-closing in Travis config

Commit:
Add server-tools in Travis config

Commit:
Inherit of product_id_change on product_id field, for better usability of the rental features

JordiBForgeFlow commented on sale_rental/README.rst:
```diff
@@ -0,0 +1,79 @@
+
+In the menu *Warehouse > Configuration > Warehouses*, on the form view
+of the warehouse, in the *Technical Information* tab, you will see two
+additionnal stock locations: *Rental In* (stock of products to rent) and
```
Replace "additionnal" with "additional".

Commit:
You can now write directly in the "number_of_days" field

Commit:
Fix flake8 warnings

Commit:
Fix spelling mistake notified by @jbeficent

max3903 commented on .travis.yml:
```diff
@@ -27,6 +27,8 @@ virtualenv:
 install:
   - git clone https://github.com/OCA/maintainer-quality-tools.git ${HOME}/maintainer-quality-tools
+  - git clone https://github.com/OCA/account-closing.git ${HOME}/account-closing
+  - git clone https://github.com/OCA/server-tools.git ${HOME}/server-tools
```
Those repositories should be added to oca_dependencies.txt and don't need to be added here, I think.

bguillot commented on sale_rental/i18n/fr.po:
```diff
@@ -0,0 +1,369 @@
+#. module: sale_rental
+#: help:sale.rental,end_date:0
+msgid "End Date of the Rental, taking into account all the extensions sold to the customer."
+msgstr "End Date of the Rental, taking into account all the extensions sold to the customer."
```
Missing translation

bguillot commented on sale_rental/i18n/fr.po:
```diff
@@ -0,0 +1,369 @@
+#: code:addons/sale_rental/rental.py:406
+#, python-format
+msgid "The Rental Service of the Rental Extension you just selected is '%s' and it's not the same as the Product currently selected in this Sale Order Line."
+msgstr "The Rental Service of the Rental Extension you just selected is '%s' and it's not the same as the Product currently selected in this Sale Order Line."
```
Missing translation

bguillot commented on sale_rental/i18n/fr.po:
```diff
@@ -0,0 +1,369 @@
+#. module: sale_rental
+#: model:stock.location,name:sale_rental.rental_out_stock_location
+msgid "Rental Out"
+msgstr "Rental Out"
```
Missing translations

bguillot commented on sale_rental/i18n/fr.po:
```diff
@@ -0,0 +1,369 @@
+#: code:addons/sale_rental/rental.py:59
+#, python-format
+msgid "The rental product '%s' must have the option ''Must Have Start and End Dates' checked."
+msgstr "The rental product '%s' must have the option ''Must Have Start and End Dates' checked."
```
Missing translation

Commit:
Move dependencies to other OCA projects to oca_dependencies.txt

Commit:
Update POT file and FR translation

Fix translated string in code

bguillot commented:
Thanks @alexis-via 

:+1:

JordiBForgeFlow commented on sale_rental/wizard/create_rental_product.py:
```diff
@@ -0,0 +1,93 @@
+            'Wrong underlying model, should be product.product'
+        hw_product = self.env['product.product'].browse(
+            self.env.context['active_id'])
+        return _('RENT-%s') % hw_product.default_code
```
The stockable product that is used as a reference to create the service rental could have the default code set as empty. In that case the current code sets "RENT-false". I suggest to add the default code for the created rental service only if the associated product has a default_code set.

JordiBForgeFlow replied:
I can see that the default code is a required field in the wizard. Why is that, if the default_code field is not a required field at all in the product? What if the original product has no default code?

JordiBForgeFlow commented on sale_rental/rental.py:
```diff
@@ -0,0 +1,636 @@
+    extension_rental_id = fields.Many2one(
+        'sale.rental', string='Rental to Extend')
+    rental_qty = fields.Float(
+        string='Rental Quantity', digits=dp.get_precision('Product UoS'))
```
I find confusing to see the following fields in the sale order line:
Number of Days
Rental Quantity
Quantity

I suggest that adding a tooltip to the field 'Rental Quantity' will help the user to clarify what to enter here.
For example:
help = "Indicate the number of items that will be delivered for rental."

JordiBForgeFlow commented on sale_rental/wizard/create_rental_product.py:
```diff
@@ -0,0 +1,93 @@
+        assert hw_product_id, 'Active ID is not set'
+        pp_obj = self.env['product.product']
+        day_uom_id = self.env.ref('product.product_uom_day').id
+
```
I suggest copying the image from the hw_product to the new rental service.

Add:
hw_product = self.env['product.product'].browse(
             hw_product_id)

JordiBForgeFlow commented on sale_rental/wizard/create_rental_product.py:
```diff
@@ -0,0 +1,93 @@
+            'default_code': self.default_code,
+            'rented_product_id': hw_product_id,
+            'must_have_dates': True,
+            'categ_id': self.categ_id.id,
```
To copy the image, add:
'image': hw_product.image,

alexis-via committed:
Take into account the remarks of Jordi Ballester

Add option to copy image from product to rental service
default_code should not be a required field in rental service wizard
Add help message on rental_qty field

alexis-via commented:
@jbeficent I took into account your remarks. It deserves a :+1:  :)

alexis-via commented:
@jbeficent Any other remarks ? We need a second positive review to merge this.

JordiBForgeFlow commented:
I have some work in progress on tests and spanish translations. Will submit a separate PR.

Another remark, I tested with module 'stock_quant_manual_assign' from @pedrobaeza and it allows you to manually select the lot/serial that you want to rent. This is really a value added function, in cases where you rent serialized equipment.

JordiBForgeFlow commented:
+1 to merge. Good work @alexis-via

sebastienbeau commented:
good work :+1:
